std/typecons.d

```diff
@@ -2675,6 +2675,11 @@ Params:
         assert(e != 12);
     }
 
+    size_t toHash() const @safe nothrow
+    {
+        return _isNull ? 0 : typeid(T).getHash(&_value.payload);
```

Comment:
When we reach release 2.082 ` typeid(T).getHash(&_value.payload)` could be replaced with `.hashOf(_value.payload)` to avoid looking up at runtime information known at compile time and to enable inlining.

Reply:
Oh, nice!

Reply:
Ah! .hashOf is @system...

Reply:
>`toHash` is not marked as `const` in `Object`.
Yeah. It's a mess.

Reply:
Casting away `const` is what `typeid(Object).getHash(&_value.payload)` does too. But because it is a `@trusted` function there's no warning when using it in `@safe` code.

[object.d#L1700-1716](https://github.com/dlang/druntime/blob/430585650930797369eeb5b3b142faba10572f10/src/object.d#L1700-L1716):
```d
class TypeInfo_Class : TypeInfo
{
    override string toString() const { return info.name; }

    override bool opEquals(Object o)
    {
        if (this is o)
            return true;
        auto c = cast(const TypeInfo_Class)o;
        return c && this.info.name == c.info.name;
    }

    override size_t getHash(scope const void* p) @trusted const
    {
        auto o = *cast(Object*)p;
        return o ? o.toHash() : 0;
    }
```

Reply:
@dkgroot I'm not familiar with the terminology widening/narrowing (more with upcast/downcast). Upcasts are implicit and fine, forced downcasts from `void*` or unrelated types are not safe for obvious reasons (incompatible layouts etc). So you're asking about casting a reference to base class to a reference to derived class? That shouldn't be unsafe in D because it returns null when the cast doesn't work.

Reply:
@andralex :
My question was partially unrelated to this PR, more a general dlang quesion (Maybe asking inside a PR Review is not the right place). In this PR the issue is centered around casting away const, which i think could/should be a considered allowable inside a @safe block.

The definition of "widening/narrowing" cast explained in the question above, they are similar to up/down casts but mostly pertain to class inheritance.

Reply:
@dkgroot 
Type qualifier casts are orthogonal to class widening/narrowing casts and are always unsafe when you're stripping them off (`const`/`immitable` -> mutable, `shared` -> non-`shared`). `immutable` is implicitly convertible to `const` on the premise that casting `const` away is undefined behavior. A `const` reference may actually point to an `immuable` object stored in read-only memory, so you can imagine what sort of damage mutating such object may cause.

Reply:
@FeepingCreature the `@safe`-ty of this method should depend on whether `.hashOf(_value.payload)` is safe - `hashOf` will be our source of truth. So you can ahead and use it, by removing the expicit `@safe` attribute and letting the compiler infer the safety.

Reply:
How would it be inferred? toHash is a regular function.

Reply:
update: As far as I can see, `.toHash` simply does not work at all, since the class `hashOf` in druntime is itself defined `@safe` but does an unsafe operation on line 615.

Reply:
>How would it be inferred? toHash is a regular function. Inference only happens for template functions.

Inferrence happens for `Nullable(T).toHash` because `Nullable` is templated. For instance:

https://run.dlang.io/is/0pWg2Z
```d
import std.stdio;

struct Nullable(T)
{
    T val;
    bool isNull;
    alias val this;
    
    this(A...)(auto ref A args) if (is(typeof(T(A)) : A)) { isNull = false; val = T(A); }
    
	size_t toHash() const nothrow
    {
        return isNull ? 0 : .hashOf(val);
    }
}

void main() @safe
{
    writeln(Nullable!int.init.toHash());
}
```